Papyrus/URLSession+Papyrus.swift

```diff
@@ -111,3 +111,48 @@ extension URLRequest: Request {
         set { allHTTPHeaderFields = newValue }
     }
 }
+
+// MARK: Retry Interceptor
+
+/// A `RetryInterceptor` that conforms to the `Interceptor` protocol.
+/// This interceptor will retry a request up to a specified number of times
+/// if it fails with a status code in the range 500-599.
+///
+/// - Parameters:
+///   - maxRetryCount: The maximum number of retry attempts. Default is 3.
+///   - initialBackoff: The initial backoff time in seconds before the first retry. Default is 2 seconds.
+///   - exponentialBackoff: The multiplier to apply to the backoff time after each retry. Default is 2.
+public struct RetryInterceptor: Interceptor {
+    private let maxRetryCount: Int
+    private let initialBackoff: UInt64
+    private let exponentialBackoff: UInt64
+
+    public init(maxRetryCount: Int = 3, initialBackoff: UInt64 = 2, exponentialBackoff: UInt64 = 2) {
+        self.maxRetryCount = maxRetryCount
+        self.initialBackoff = initialBackoff
+        self.exponentialBackoff = exponentialBackoff
+    }
+
+    public func intercept(req: Request, next: Next) async throws -> Response {
+        var response: Response
+        var retryCount = 0
+        var backoff = initialBackoff
+
+        repeat {
+            do {
+                response = try await next(req)
+                if let statusCode = response.statusCode, (500...599).contains(statusCode) {
```

Comment:
Probably if we implement just retry strategy - `needRetry` could be an external logic instead of hardcoded one.
So we'll be able to check anything inside response to decide if retry needed.

Reply:
Yep totally agree - this PR was actually made by a bot. The GitHub copilot version (other PR) does what you suggest and I think is a much better option.